Arm Subsystem Completed with New State Structure

rohitchawla28 opened the pull request:
## Arm Subsystem General Structure

- Chooses between 3 states to perform particular actions respective to the state
   - ArmState::DISABLED
   - ArmState::DISENGAGE
   - ArmState::MANUAL
- Runs all gather-info/output actions in Periodic() loop (every 20 ms)
- GetInstance() allows other classes to utilize the Arm subsystem without creating new memory each time called
- x struct variable (declared as state in .h) used to keep track of loop-to-loop information
   - current state
   - current power
   - disengage boolean indicator
   - disengage helper variables
      - step 1 start time and step 2 start time of disengage process
   - frc::Timer object
   - current FPGA time


## Initializing the Arm Subsystem

- Takes in pointer to XboxController object as constructor parameter to allow for reading of gamepad
- XboxController member variable is set to pointer parameter in initializer list
- InitArm() initializes Talon SRX motors for arm and resets other state struct variables
- setDefaultState() is called in Robot.cpp in DisabledInit period to ensure Arm is in ArmState::DISABLED


## Periodic Loop

- Calls AssessInputs() and AssignOutputs() methods here


### AssessInputs()

- Reads input values and assigns the Arm to one of the states
- Uses state.disengage boolean to decide if Arm should be in auto disengage process
- Uses operator gamepad left stick Y axis to decide if Arm should be in tele-op manual mode
* Disabled state is controlled outside of Periodic() loop (setDefaultState())


### AssignOutputs()

- Reads states and sets desired power for Arm motors
- Uses state.arm_state to see which state Arm subsystem is currently in

#### Disengage Process

- Process is carried out at start of auto period to bring arm down from pin to robot frame
- Uses -1 values for step1/step2 start time variables to check if step1/step2 have been started
- Uses state.timer.getFPGATimestamp() for accurate reading at beginning of each loop iteration

**Time values have not been tested yet

1. Raise arm slightly out of pin (upwards power for very short time)
2. Lower arm carefully until arm reaches frame (downwards power for little longer time)

#### Manual Arm Control

- State meant for full tele-op period

Arm Manual Control Checklist

1. If joystick hasn't moved, set 0
2. If joystick is pulled down to any amount less than -0.05, set -0.5
3. If joystick pushed up between 0.05 and 0.85, set -0.042
4. Else (pushed up to full) set 0.1

rohitchawla28 committed:
completed arm subsystem with new state structure

rohitchawla28 committed:
fixed extra/wrong std::abs calls in manual arm control

mray190 commented on Competition/src/main/cpp/Arm.cpp:
```diff
@@ -0,0 +1,98 @@
+    return instance;
+}
+
+void Arm::Periodic() {
```
You do not need this function. The super class does this for you. Can remove the definition in the header file, and delete this function from the implementation

rohitchawla28 replied:
Okay that makes sense.

mray190 commented on Competition/src/main/cpp/Arm.cpp:
```diff
@@ -0,0 +1,98 @@
+    InitArm();
+}
+
+Arm& Arm::GetInstance() {
```
You do not need this function. The super class does this for you. Can remove the definition in the header file, and delete this function from the implementation

mray190 commented on Competition/src/main/cpp/Arm.cpp:
```diff
@@ -0,0 +1,98 @@
+#include "Arm.h"
+
+Arm::Arm() : armMtrLeft{ArmConstants::TALON_ID_LEFT_ARM}, 
```
Not passing in a controller pointer would leave the controller null, and therefore seg fault later in the code when trying to utilize the controller member variable.

Therefore, do not allow the user to use a constructor that doesn't pass in a controller pointer. Delete this implementation and delete the definition in the header file

rohitchawla28 replied:
Okay, the only reason I had this is because when I had the GetInstance() method in, it would give me an error saying there was no default constructor for when it did "static Arm instance" but I guess that really isn't a good solution to the problem at all.

mray190 replied:
Ahh that's a good point. I'll need to think about that, cause yeah that's a problem

mray190 commented on Competition/src/main/cpp/Arm.cpp:
```diff
@@ -0,0 +1,98 @@
+    armMtrLeft.SetInverted(false);
+    armMtrRight.SetInverted(false);
+
+    state.disengage = true;
```
This should be set in the setDefaultState function. Same with the step timers

mray190 commented on Competition/src/main/cpp/Arm.cpp:
```diff
@@ -0,0 +1,98 @@
+}
+
+void Arm::assessInputs() {
+    if (state.disengage) {
```
One of the artificial rules we set was that you could not read the state variable from within the assessInputs function. This is to make sure ALL decisions are made within the assignOutputs function.

Therefore this if statement should be moved into assignOutputs. The else case is perfect though

rohitchawla28 replied:
Just confirming, so having this if statement just to set state.arm_state in AssignOutputs() is fine then? It would check the boolean for disengage and assign the state if true and then jump into the other AssignOutputs() logic after.

mray190 replied:
Yup that works!

mray190 commented on Competition/src/main/cpp/Arm.cpp:
```diff
@@ -0,0 +1,98 @@
+            state.current_power = 0;
+
+            // reset code
+            // should reset code be in setDefaultState()?
```
reset code can be in its own function - perhaps resetState.

Then setDefaultState() can call resetState() among other things

rohitchawla28 replied:
Hmm, yeah need to figure out when I am changing the disengage boolean. Because it should be set to true at the start of auto, and should be set to false immediately after the process. But we wouldn't want it to set to true in setDefaultState/DisabledInit right? Because if we are just turning on the robot for tele-op practice or something, we don't want to start disengaging. So maybe AutonomousInit is a good place for disengage to set true, but setDefaultState can have it set false still

mray190 commented on Competition/src/main/cpp/Arm.cpp:
```diff
@@ -0,0 +1,98 @@
+        }
+    }
+    else if (state.arm_state == ArmState::MANUAL) {
+        if (std::abs(operator_controller->GetY(frc::GenericHID::kLeftHand)) <= 0.05) {
```
Reading from the joystick should occur in assessInputs and store into a state.joystick variable.

All the following if statements should then reference state.joystick

mray190 reviewed (changes requested):
I haven't evaluated or stepped through any of the logic in assignOutputs - I can do that once some of these other things are fixed.

For your first attempt, this looks great!! Great documentation in the PR summary, header file for Arm looks really good, and good work including the CTRE library

rohitchawla28 committed:
PR review fixes - removed unnecess arm methods, added joystick to state

mray190 committed:
Provided example on how to pass in gamepad objects

Member function in the Arm object to set the controller
Global constants for gamepad USB ports
Set the controller for the arm in the ConfigureButtonBindings()

rohitchawla28 committed:
set controller method and fixed capitalization/spacing

rohitchawla28 committed:
added reset state after disengage process, made setDisengage function

mray190 commented:
@rohitchawla28 We will merge this tomorrow during our meeting (or class).
I want to show all of you how to rebase
